### ISSUE TITLE
A Visual Intro to NumPy and Data Representation  Jay Alammar  Visualizing machine learning one concept at a time.

### ISSUE THREAD

**Issue by guevara** [guevara/read-it-later#9398]
A Visual Intro to NumPy and Data Representation – Jay Alammar – Visualizing machine learning one concept at a time.<br>
<br>
https://ift.tt/bnStZPg<br>
<br>
Jay Alammar<br>
<br>
<div><div>
<p><span>Discussions:
<a href="https://news.ycombinator.com/item?id=20282985">Hacker News (366 points, 21 comments)</a>, <a href="https://www.reddit.com/r/MachineLearning/comments/c5nc89/p_a_visual_intro_to_numpy_and_data_representation/">Reddit r/MachineLearning (256 points, 18 comments)</a>
</span>
<br>
<span>Translations: <a href="http://www.junphy.com/wordpress/index.php/2019/10/24/visual-numpy/">Chinese 1</a>, <a href="https://github.com/kevingo/blog/blob/master/ML/visual-numpy.md">Chinese 2</a>, <a href="https://note.mu/sayajewels/n/n95edaedb0fc5">Japanese</a>, <a href="https://chloamme.github.io/2021/12/20/visual-numpy-korean.html">Korean</a></span></p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array.png">
<br>
</div>
<p>The <a href="https://www.numpy.org/">NumPy</a> package is the workhorse of data analysis, machine learning, and scientific computing in the python ecosystem. It vastly simplifies manipulating and crunching vectors and matrices. Some of python’s leading package rely on NumPy as a fundamental piece of their infrastructure (examples include scikit-learn, SciPy, pandas, and tensorflow). Beyond the ability to slice and dice numeric data, mastering numpy will give you an edge when dealing and debugging with advanced usecases in these libraries.</p>
<p>In this post, we’ll look at some of the main ways to use NumPy and how it can represent different types of data (tables, images, text…etc) before we can serve them to machine learning models.</p>

<h2>Creating Arrays</h2>
<p>We can create a NumPy array (a.k.a. the mighty <a href="https://docs.scipy.org/doc/numpy/reference/arrays.ndarray.html">ndarray</a>) by passing a python list to it and using ` np.array()`. In this case, python creates the array we can see on the right here:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/create-numpy-array-1.png">
<br>
</div>
<p>There are often cases when we want NumPy to initialize the values of the array for us. NumPy provides methods like ones(), zeros(), and random.random() for these cases. We just pass them the number of elements we want it to generate:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/create-numpy-array-ones-zeros-random.png">
<br>
</div>
<p>Once we’ve created our arrays, we can start to manipulate them in interesting ways.</p>
<h2>Array Arithmetic</h2>
<p>Let’s create two NumPy arrays to showcase their usefulness. We’ll call them <code>data</code> and <code>ones</code>:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-arrays-example-1.png">
<br>
</div>
<p>Adding them up position-wise (i.e. adding the values of each row) is as simple as typing <code>data + ones</code>:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-arrays-adding-1.png">
<br>
</div>
<p>When I started learning such tools, I found it refreshing that an abstraction like this makes me not have to program such a calculation in loops. It’s a wonderful abstraction that allows you to think about problems at a higher level.</p>
<p>And it’s not only addition that we can do this way:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array-subtract-multiply-divide.png">
<br>
</div>
<p>There are often cases when we want to carry out an operation between an array and a single number (we can also call this an operation between a vector and a scalar). Say, for example, our array represents distance in miles, and we want to convert it to kilometers. We simply say <code>data * 1.6</code>:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array-broadcast.png">
<br>
</div>
<p>See how NumPy understood that operation to mean that the multiplication should happen with each cell? That concept is called <em>broadcasting</em>, and it’s very useful.</p>
<h2>Indexing</h2>
<p>We can index and slice NumPy arrays in all the ways we can slice python lists:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array-slice.png">
<br>
</div>
<h2>Aggregation</h2>
<p>Additional benefits NumPy gives us are aggregation functions:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array-aggregation.png">
<br>
</div>
<p>In addition to <code>min</code>, <code>max</code>, and <code>sum</code>, you get all the greats like <code>mean</code> to get the average, <code>prod</code> to get the result of multiplying all the elements together, <code>std</code> to get standard deviation, and <a href="https://jakevdp.github.io/PythonDataScienceHandbook/02.04-computation-on-arrays-aggregates.html">plenty of others</a>.</p>
<h2>In more dimensions</h2>
<p>All the examples we’ve looked at deal with vectors in one dimension. A key part of the beauty of NumPy is its ability to apply everything we’ve looked at so far to any number of dimensions.</p>
<h3>Creating Matrices</h3>
<p>We can pass python lists of lists in the following shape to have NumPy create a matrix to represent them:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-array-create-2d.png">
<br>
</div>
<p>We can also use the same methods we mentioned above (<code>ones()</code>, <code>zeros()</code>, and <code>random.random()</code>) as long as we give them a tuple describing the dimensions of the matrix we are creating:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-ones-zeros-random.png">
<br>
</div>
<h3>Matrix Arithmetic</h3>
<p>We can add and multiply matrices using arithmetic operators (<code>+-*/</code>) if the two matrices are the same size. NumPy handles those as position-wise operations:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-arithmetic.png">
<br>
</div>
<p>We can get away with doing these arithmetic operations on matrices of different size only if the different dimension is one (e.g. the matrix has only one column or one row), in which case NumPy uses its broadcast rules for that operation:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-broadcast.png">
<br>
</div>
<h3>Dot Product</h3>
<p>A key distinction to make with arithmetic is the case of <a href="https://www.mathsisfun.com/algebra/matrix-multiplying.html">matrix multiplication</a> using the dot product. NumPy gives every matrix a <code>dot()</code> method we can use to carry-out dot product operations with other matrices:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-dot-product-1.png">
<br>
</div>
<p>I’ve added matrix dimensions at the bottom of this figure to stress that the two matrices have to have the same dimension on the side they face each other with. You can visualize this operation as looking like this:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-dot-product-2.png">
<br>
</div>
<h3>Matrix Indexing</h3>
<p>Indexing and slicing operations become even more useful when we’re manipulating matrices:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-indexing.png">
<br>
</div>
<h3>Matrix Aggregation</h3>
<p>We can aggregate matrices the same way we aggregated vectors:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-aggregation-1.png">
<br>
</div>
<p>Not only can we aggregate all the values in a matrix, but we can also aggregate across the rows or columns by using the <code>axis</code> parameter:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-matrix-aggregation-4.png">
<br>
</div>
<h2>Transposing and Reshaping</h2>
<p>A common need when dealing with matrices is the need to rotate them. This is often the case when we need to take the dot product of two matrices and need to align the dimension they share. NumPy arrays have a convenient property called <code>T</code> to get the transpose of a matrix:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-transpose.png">
<br>
</div>
<p>In more advanced use case, you may find yourself needing to switch the dimensions of a certain matrix. This is often the case in machine learning applications where a certain model expects a certain shape for the inputs that is different from your dataset. NumPy’s <code>reshape()</code> method is useful in these cases. You just pass it the new dimensions you want for the matrix. You can pass -1 for a dimension and NumPy can infer the correct dimension based on your matrix:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-reshape.png">
<br>
</div>
<h2>Yet More Dimensions</h2>
<p>NumPy can do everything we’ve mentioned in any number of dimensions. Its central data structure is called ndarray (N-Dimensional Array) for a reason.</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-3d-array.png">
<br>
</div>
<p>In a lot of ways, dealing with a new dimension is just adding a comma to the parameters of a NumPy function:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-3d-array-creation.png">
<br>
</div>
<p>Note: Keep in mind that when you print a 3-dimensional NumPy array, the text output visualizes the array differently than shown here. NumPy’s order for printing n-dimensional arrays is that the last axis is looped over the fastest, while the first is the slowest. Which means that <code>np.ones((4,3,2))</code> would be printed as:</p>
<div><div><pre><code><span>array</span><span>([[[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>]],</span>

       <span>[[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>]],</span>

       <span>[[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>]],</span>

       <span>[[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>],</span>
        <span>[</span><span>1.</span><span>,</span> <span>1.</span><span>]]])</span>
</code></pre></div></div>
<h2>Practical Usage</h2>
<p>And now for the payoff. Here are some examples of the useful things NumPy will help you through.</p>
<h3>Formulas</h3>
<p>Implementing mathematical formulas that work on matrices and vectors is a key use case to consider NumPy for. It’s why NumPy is the darling of the scientific python community. For example, consider the mean square error formula that is central to supervised machine learning models tackling regression problems:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/mean-square-error-formula.png">
<br>
</div>
<p>Implementing this is a breeze in NumPy:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-mean-square-error-formula.png">
<br>
</div>
<p>The beauty of this is that numpy does not care if <code>predictions</code> and <code>labels</code> contain one or a thousand values (as long as they’re both the same size). We can walk through an example stepping sequentially through the four operations in that line of code:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-mse-1.png">
<br>
</div>
<p>Both the predictions and labels vectors contain three values. Which means n has a value of three. After we carry out the subtraction, we end up with the values looking like this:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-mse-2.png">
<br>
</div>
<p>Then we can square the values in the vector:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-mse-3.png">
<br>
</div>
<p>Now we sum these values:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-mse-4.png">
<br>
</div>
<p>Which results in the error value for that prediction and a score for the quality of the model.</p>
<h3>Data Representation</h3>
<p>Think of all the data types you’ll need to crunch and build models around (spreadsheets, images, audio…etc). So many of them are perfectly suited for representation in an n-dimensional array:</p>
<h4>Tables and Spreadsheets</h4>
<ul>
<li>A spreadsheet or a table of values is a two dimensional matrix. Each sheet in a spreadsheet can be its own variable. The most popular abstraction in python for those is the <a href="https://jalammar.github.io/gentle-visual-intro-to-data-analysis-python-pandas/">pandas dataframe</a>, which actually uses NumPy and builds on top of it.</li>
</ul>
<div>
<img src="https://jalammar.github.io/images/pandas-intro/0%20excel-to-pandas.png">
<br>
</div>
<h4>Audio and Timeseries</h4>
<ul>
<li>An audio file is a one-dimensional array of samples. Each sample is a number representing a tiny chunk of the audio signal. CD-quality audio may have 44,100 samples per second and each sample is an integer between -32767 and 32768. Meaning if you have a ten-seconds WAVE file of CD-quality, you can load it in a NumPy array with length 10 * 44,100 = 441,000 samples. Want to extract the first second of audio? simply load the file into a NumPy array that we’ll call <code>audio</code>, and get <code>audio[:44100]</code>.</li>
</ul>
<p>Here’s a look at a slice of an audio file:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-audio.png">
<br>
</div>
<p>The same goes for time-series data (for example, the price of a stock over time).</p>
<h4>Images</h4>
<ul>
<li>
<p>An image is a matrix of pixels of size (height x width).</p>
<ul>
<li>If the image is black and white (a.k.a. grayscale), each pixel can be represented by a single number (commonly between 0 (black) and 255 (white)). Want to crop the top left 10 x 10 pixel part of the image? Just tell NumPy to get you <code>image[:10,:10]</code>.</li>
</ul>
</li>
</ul>
<p>Here’s a look at a slice of an image file:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-grayscale-image.png">
<br>
</div>
<ul>
<li>
<p>If the image is colored, then each pixel is represented by three numbers - a value for each of red, green, and blue. In that case we need a 3rd dimension (because each cell can only contain one number). So a colored image is represented by an ndarray of dimensions: (height x width x 3).</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-color-image.png">
<br>
</div>
</li>
</ul>
<h4>Language</h4>
<p>If we’re dealing with text, the story is a little different. The numeric representation of text requires a step of building a vocabulary (an inventory of all the unique words the model knows) and an <a href="https://jalammar.github.io/illustrated-word2vec/">embedding step</a>. Let us see the steps of numerically representing this (translated) quote by an ancient spirit:</p>
<p>“Have the bards who preceded me left any theme unsung?”</p>
<p>A model needs to look at a large amount of text before it can numerically represent the anxious words of this warrior poet. We can proceed to have it process a <a href="http://mattmahoney.net/dc/textdata.html">small dataset</a> and use it to build a vocabulary (of 71,290 words):</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-nlp-vocabulary.png">
<br>
</div>
<p>The sentence can then be broken into an array of tokens (words or parts of words based on common rules):</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-nlp-tokenization.png">
<br>
</div>
<p>We then replace each word by its id in the vocabulary table:</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-nlp-ids.png">
<br>
</div>
<p>These ids still don’t provide much information value to a model. So before feeding a sequence of words to a model, the tokens/words need to be replaced with their embeddings (50 dimension <a href="https://jalammar.github.io/illustrated-word2vec/">word2vec embedding</a> in this case):</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-nlp-embeddings.png">
<br>
</div>
<p>You can see that this NumPy array has the dimensions [embedding_dimension x sequence_length]. In practice these would be the other way around, but I’m presenting it this way for visual consistency. For performance reasons, deep learning models tend to preserve the first dimension for batch size (because the model can be trained faster if multiple examples are trained in parallel). This is a clear case where <code>reshape()</code> becomes super useful. A model like <a href="https://jalammar.github.io/illustrated-bert/">BERT</a>, for example, would expect its inputs in the shape: [batch_size, sequence_length, embedding_size].</p>
<div>
<img src="https://jalammar.github.io/images/numpy/numpy-nlp-bert-shape.png">
<br>
</div>
<p>This is now a numeric volume that a model can crunch and do useful things with. I left the other rows empty, but they’d be filled with other examples for the model to train on (or predict).</p>
<p>(It turned out the <a href="https://en.wikisource.org/wiki/The_Poem_of_Antara">poet’s words</a> in our example were immortalized more so than those of the other poets which trigger his anxieties. Born a slave owned by his father, <a href="https://en.wikipedia.org/wiki/Antarah_ibn_Shaddad">Antarah’s</a> valor and command of language gained him his freedom and the mythical status of having his poem as one of <a href="https://en.wikipedia.org/wiki/Mu%27allaqat">seven poems suspended in the kaaba</a> in pre-Islamic Arabia).</p>
</div></div><br>
<br>
<br>
<br>
<br>
<br>
via jalammar.github.io https://ift.tt/GQiyEaX<br>
<br>
June 2, 2023 at 04:07PM
